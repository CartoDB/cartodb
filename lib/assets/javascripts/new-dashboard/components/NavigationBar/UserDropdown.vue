<template>
  <div class="navbar-dropdown" :class="{ 'is-open': open }">
    <ul class="navbar-dropdown-container">
      <li class="navbar-dropdown-profile">
        <div class="avatar-container">
          <img :src="userModel.avatar_url">
        </div>
        <div class="navbar-dropdown-userInfo">
          <p class="text is-semibold is-caption">{{userModel.username}}</p>
          <p class="text is-small">{{userModel.email}}</p>
        </div>
      </li>
      <li class="navbar-dropdown-iconLink">
        <div class="icon-container">
          <img svg-inline src="../../assets/icons/navbar/dropdown/settings.svg"/>
        </div>
        <a :href="`${ baseUrl }/profile`" class="text is-semibold is-caption is-txtGrey" @click="linkClicked">{{ $t('UserDropdown.settings') }}</a>
      </li>
      <li class="navbar-dropdown-iconLink" v-if="isOrganizationAdmin">
        <div class="icon-container">
          <img svg-inline src="../../assets/icons/navbar/dropdown/organization.svg"/>
        </div>
        <a :href="`${ baseUrl }/organization`" class="text is-semibold is-caption is-txtGrey" @click="linkClicked">{{ $t('UserDropdown.organizationSettings') }}</a>
      </li>
      <li @click="linkClicked" class="navbar-dropdown-iconLink">
        <div class="icon-container">
          <img svg-inline src="../../assets/icons/navbar/dropdown/api-keys.svg"/>
        </div>
        <a :href="`${ baseUrl }/your_apps`" class="text is-semibold is-caption is-txtGrey" @click="linkClicked">{{ $t('UserDropdown.apiKeys') }}</a>
      </li>
      <li class="navbar-dropdown-iconLink">
        <div class="icon-container">
          <img svg-inline src="../../assets/icons/navbar/dropdown/profile.svg"/>
        </div>
        <a :href="`${ baseUrl }/me`" class="text is-semibold is-caption is-txtGrey" target="_blank" @click="linkClicked" >{{ $t('UserDropdown.publicProfile') }}</a>
      </li>
      <li class="navbar-dropdown-iconLink">
        <div class="icon-container">
          <img svg-inline src="../../assets/icons/navbar/dropdown/notifications.svg"/>
        </div>
        <router-link :to="{ name: 'notifications' }" class="text is-semibold is-caption is-txtGrey" @click.native="linkClicked">
          {{ $t('UserDropdown.notifications') }}
        </router-link>
        <span v-if="notificationsCount > 0" class="notification-number text is-semibold is-small is-txtGrey">{{notificationsCount}}</span>
      </li>
    </ul>
    <a :href="`${ baseUrl }/logout`" class="navbar-dropdown-bottom" @click="linkClicked">
      <div class="arrow"><img svg-inline src="../../assets/icons/navbar/dropdown/arrow-right.svg" /></div>
      <div class="title is-small is-txtAlert">{{ $t('UserDropdown.logout') }}</div>
    </a>
  </div>
</template>

<script>
export default {
  name: 'UserDropdown',
  props: {
    baseUrl: { type: String, default: '' },
    open: { type: Boolean, default: false },
    userModel: Object,
    notificationsCount: Number
  },
  computed: {
    isOrganizationAdmin () {
      return this.userModel.org_admin;
    }
  },
  methods: {
    linkClicked () {
      this.$emit('linkClick');
    }
  }
};
</script>
<style scoped lang="scss">
@import 'stylesheets/new-dashboard/variables';

.navbar-dropdown {
  visibility: hidden;
  position: absolute;
  right: 0;
  width: 360px;
  margin-top: 18px;
  transition: all 0.25s linear;
  border-radius: 4px;
  opacity: 0;
  box-shadow: 0 2px 8px 0 rgba(0, 0, 0, 0.16);
  pointer-events: none;

  &::before {
    content: "";
    position: absolute;
    right: 0;
    width: 20px;
    height: 20px;
    transform: rotate(45deg) translateX(-50%);
    transform-origin: center center;
    border-radius: 4px;
    background-color: $white;
    box-shadow: 0 2px 8px 0 rgba(0, 0, 0, 0.16);
  }

  &.is-open {
    visibility: visible;
    opacity: 1;
    pointer-events: initial;
  }
}

.navbar-dropdown-profile {
  display: flex;
  align-items: center;
  margin-bottom: 36px;

  .avatar-container {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 36px;
    height: 36px;
    margin-right: 12px;
    overflow: hidden;
    border-radius: 50%;

    img {
      width: 100%;
      object-fit: cover;
    }
  }
}

.navbar-dropdown-container {
  position: relative;
  padding: 28px 24px 36px;
  border-radius: 4px 4px 0 0;
  background-color: $white;
}

.navbar-dropdown-iconLink {
  display: flex;
  margin-bottom: 20px;

  .icon-container {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 24px;
    height: 24px;
    margin-right: 8px;
  }

  &:last-of-type {
    margin-bottom: 0;
  }
}

.navbar-dropdown-bottom {
  display: flex;
  align-items: baseline;
  padding: 24px 30px;
  border-radius: 0 0 4px 4px;
  background-color: $softblue;

  &:hover {
    text-decoration: none;
  }

  & .arrow {
    margin-right: 8px;

    svg {
      vertical-align: middle;
    }
  }

  .title {
    letter-spacing: 1px;
  }
}

.notification-number {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 20px;
  height: 20px;
  margin-top: 2px;
  margin-left: 24px;
  border-radius: 50%;
  background-color: $notification;
}
</style>
