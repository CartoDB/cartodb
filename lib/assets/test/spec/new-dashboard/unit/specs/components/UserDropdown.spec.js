import { mount } from '@vue/test-utils';
import UserDropdown from 'new-dashboard/components/NavigationBar/UserDropdown';

const $t = key => key;

describe('UserDropdown.vue', () => {
  let user;
  beforeEach(() => {
    user = {
      avatar_url: '//cartodb-libs.global.ssl.fastly.net/cartodbui/assets/unversioned/images/avatars/avatar_ghost_yellow.png',
      username: 'Test user',
      email: 'user@test.com'
    };
  });

  it('should render dropdown open with notifications badge', () => {
    const userDropdown = mount(UserDropdown, {
      propsData: {
        userModel: user,
        notificationsCount: 2
      },
      mocks: { $t }
    });
    expect(userDropdown).toMatchSnapshot();
  });

  it('should render dropdown closed', () => {
    const userDropdown = mount(UserDropdown, {
      propsData: {
        open: true,
        userModel: user,
        notificationsCount: 0
      },
      mocks: { $t }
    });

    expect(userDropdown).toMatchSnapshot();
  });

  it('should render organization link if user is org_admin', () => {
    const overridenUser = {
      ...user,
      org_admin: true
    };

    const userDropdown = mount(UserDropdown, {
      propsData: {
        open: true,
        userModel: overridenUser,
        notificationsCount: 0
      },
      mocks: { $t }
    });

    expect(userDropdown).toMatchSnapshot();
  });
});
