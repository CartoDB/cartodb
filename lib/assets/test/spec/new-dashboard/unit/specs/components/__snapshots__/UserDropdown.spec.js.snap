// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDropdown.vue should render dropdown closed 1`] = `
<div class="navbar-dropdown is-open">
  <ul class="navbar-dropdown-container">
    <li class="navbar-dropdown-profile">
      <div class="avatar-container"><img src="//cartodb-libs.global.ssl.fastly.net/cartodbui/assets/unversioned/images/avatars/avatar_ghost_yellow.png"></div>
      <div class="navbar-dropdown-userInfo">
        <p class="text is-semibold is-caption">Test user</p>
        <p class="text is-small">user@test.com</p>
      </div>
    </li>
    <li class="navbar-dropdown-iconLink">
      <div class="icon-container"><img svg-inline="" src="../../assets/icons/navbar/dropdown/settings.svg"></div> <a href="/profile" class="text is-semibold is-caption is-txtGrey">UserDropdown.settings</a>
    </li>
    <!---->
    <li class="navbar-dropdown-iconLink">
      <div class="icon-container"><img svg-inline="" src="../../assets/icons/navbar/dropdown/api-keys.svg"></div> <a href="/your_apps" class="text is-semibold is-caption is-txtGrey">UserDropdown.apiKeys</a>
    </li>
    <li class="navbar-dropdown-iconLink">
      <div class="icon-container"><img svg-inline="" src="../../assets/icons/navbar/dropdown/profile.svg"></div> <a href="/me" target="_blank" class="text is-semibold is-caption is-txtGrey">UserDropdown.publicProfile</a>
    </li>
    <li class="navbar-dropdown-iconLink">
      <div class="icon-container"><img svg-inline="" src="../../assets/icons/navbar/dropdown/notifications.svg"></div>
      <router-link to="[object Object]" staticroute="/dashboard/notifications" class="text is-semibold is-caption is-txtGrey">
        UserDropdown.notifications
      </router-link>
      <!---->
    </li>
    <li class="navbar-dropdown-iconLink">
      <div class="icon-container"><img svg-inline="" src="../../assets/icons/navbar/dropdown/feedback.svg"></div> <a href="https://docs.google.com/forms/d/e/1FAIpQLScBQUWd-TP3Qy514DOCNg-KoLrViHijUR5giLAMS-3jmDnrPg/viewform" target="_blank" class="text is-semibold is-caption is-txtGrey">
        UserDropdown.feedback
      </a>
    </li>
  </ul> <a href="/logout" class="navbar-dropdown-bottom">
    <div class="arrow"><img src="../../assets/icons/navbar/dropdown/arrow-right.svg"></div>
    <div class="title is-small is-txtAlert">UserDropdown.logout</div>
  </a>
</div>
`;

exports[`UserDropdown.vue should render dropdown open with notifications badge 1`] = `
<div class="navbar-dropdown">
  <ul class="navbar-dropdown-container">
    <li class="navbar-dropdown-profile">
      <div class="avatar-container"><img src="//cartodb-libs.global.ssl.fastly.net/cartodbui/assets/unversioned/images/avatars/avatar_ghost_yellow.png"></div>
      <div class="navbar-dropdown-userInfo">
        <p class="text is-semibold is-caption">Test user</p>
        <p class="text is-small">user@test.com</p>
      </div>
    </li>
    <li class="navbar-dropdown-iconLink">
      <div class="icon-container"><img svg-inline="" src="../../assets/icons/navbar/dropdown/settings.svg"></div> <a href="/profile" class="text is-semibold is-caption is-txtGrey">UserDropdown.settings</a>
    </li>
    <!---->
    <li class="navbar-dropdown-iconLink">
      <div class="icon-container"><img svg-inline="" src="../../assets/icons/navbar/dropdown/api-keys.svg"></div> <a href="/your_apps" class="text is-semibold is-caption is-txtGrey">UserDropdown.apiKeys</a>
    </li>
    <li class="navbar-dropdown-iconLink">
      <div class="icon-container"><img svg-inline="" src="../../assets/icons/navbar/dropdown/profile.svg"></div> <a href="/me" target="_blank" class="text is-semibold is-caption is-txtGrey">UserDropdown.publicProfile</a>
    </li>
    <li class="navbar-dropdown-iconLink">
      <div class="icon-container"><img svg-inline="" src="../../assets/icons/navbar/dropdown/notifications.svg"></div>
      <router-link to="[object Object]" staticroute="/dashboard/notifications" class="text is-semibold is-caption is-txtGrey">
        UserDropdown.notifications
      </router-link> <span class="notification-number text is-semibold is-small is-txtGrey">2</span>
    </li>
    <li class="navbar-dropdown-iconLink">
      <div class="icon-container"><img svg-inline="" src="../../assets/icons/navbar/dropdown/feedback.svg"></div> <a href="https://docs.google.com/forms/d/e/1FAIpQLScBQUWd-TP3Qy514DOCNg-KoLrViHijUR5giLAMS-3jmDnrPg/viewform" target="_blank" class="text is-semibold is-caption is-txtGrey">
        UserDropdown.feedback
      </a>
    </li>
  </ul> <a href="/logout" class="navbar-dropdown-bottom">
    <div class="arrow"><img src="../../assets/icons/navbar/dropdown/arrow-right.svg"></div>
    <div class="title is-small is-txtAlert">UserDropdown.logout</div>
  </a>
</div>
`;

exports[`UserDropdown.vue should render organization link if user is org_admin 1`] = `
<div class="navbar-dropdown is-open">
  <ul class="navbar-dropdown-container">
    <li class="navbar-dropdown-profile">
      <div class="avatar-container"><img src="//cartodb-libs.global.ssl.fastly.net/cartodbui/assets/unversioned/images/avatars/avatar_ghost_yellow.png"></div>
      <div class="navbar-dropdown-userInfo">
        <p class="text is-semibold is-caption">Test user</p>
        <p class="text is-small">user@test.com</p>
      </div>
    </li>
    <li class="navbar-dropdown-iconLink">
      <div class="icon-container"><img svg-inline="" src="../../assets/icons/navbar/dropdown/settings.svg"></div> <a href="/profile" class="text is-semibold is-caption is-txtGrey">UserDropdown.settings</a>
    </li>
    <li class="navbar-dropdown-iconLink">
      <div class="icon-container"><img svg-inline="" src="../../assets/icons/navbar/dropdown/organization.svg"></div> <a href="/organization" class="text is-semibold is-caption is-txtGrey">UserDropdown.organizationSettings</a>
    </li>
    <li class="navbar-dropdown-iconLink">
      <div class="icon-container"><img svg-inline="" src="../../assets/icons/navbar/dropdown/api-keys.svg"></div> <a href="/your_apps" class="text is-semibold is-caption is-txtGrey">UserDropdown.apiKeys</a>
    </li>
    <li class="navbar-dropdown-iconLink">
      <div class="icon-container"><img svg-inline="" src="../../assets/icons/navbar/dropdown/profile.svg"></div> <a href="/me" target="_blank" class="text is-semibold is-caption is-txtGrey">UserDropdown.publicProfile</a>
    </li>
    <li class="navbar-dropdown-iconLink">
      <div class="icon-container"><img svg-inline="" src="../../assets/icons/navbar/dropdown/notifications.svg"></div>
      <router-link to="[object Object]" staticroute="/dashboard/notifications" class="text is-semibold is-caption is-txtGrey">
        UserDropdown.notifications
      </router-link>
      <!---->
    </li>
    <li class="navbar-dropdown-iconLink">
      <div class="icon-container"><img svg-inline="" src="../../assets/icons/navbar/dropdown/feedback.svg"></div> <a href="https://docs.google.com/forms/d/e/1FAIpQLScBQUWd-TP3Qy514DOCNg-KoLrViHijUR5giLAMS-3jmDnrPg/viewform" target="_blank" class="text is-semibold is-caption is-txtGrey">
        UserDropdown.feedback
      </a>
    </li>
  </ul> <a href="/logout" class="navbar-dropdown-bottom">
    <div class="arrow"><img src="../../assets/icons/navbar/dropdown/arrow-right.svg"></div>
    <div class="title is-small is-txtAlert">UserDropdown.logout</div>
  </a>
</div>
`;
